packages/theme/src/cli/utilities/theme-environment/hot-reload/server.ts

```diff
@@ -0,0 +1,186 @@
+import {getClientScripts, HotReloadEvent} from './client.js'
+import {render} from '../storefront-renderer.js'
+import {patchRenderingResponse} from '../proxy.js'
+import {createEventStream, defineEventHandler, getProxyRequestHeaders, getQuery, sendError, type H3Error} from 'h3'
+import {renderWarning} from '@shopify/cli-kit/node/ui'
+import {extname} from '@shopify/cli-kit/node/path'
+import EventEmitter from 'node:events'
+import type {Theme, ThemeFSEvent} from '@shopify/cli-kit/node/themes/types'
+import type {DevServerContext} from '../types.js'
+
+// --- Template Replacers ---
+
+interface TemplateWithSections {
+  sections?: {[key: string]: {type: string}}
+}
+
+const inMemoryTemplates = {} as {[key: string]: string}
+const parsedJsonTemplates = {} as {[key: string]: TemplateWithSections}
+
+/**
+ * Gets all the modified files recorded in memory for `replaceTemplates` in the API.
+ */
+export function getInMemoryTemplates() {
+  return {...inMemoryTemplates}
+}
+
+/**
+ * Watchs for file changes and updates in-memory templates, triggering
+ * HotReload if needed.
+ */
+export function setupInMemoryTemplateWatcher(ctx: DevServerContext) {
+  const handleFileDelete = ({fileKey, syncPromise}: ThemeFSEvent<'unlink'>['payload'], deleteJsonValue = true) => {
+    syncPromise
+      .then(() => {
+        // Delete memory info after syncing with the remote instance because we
+        // don't need to pass replaceTemplates anymore.
+        delete inMemoryTemplates[fileKey]
+        if (deleteJsonValue) delete parsedJsonTemplates[fileKey]
+      })
+      .catch(() => {})
+  }
+
+  const handleFileUpdate = ({fileKey, contentPromise, syncPromise}: ThemeFSEvent<'add'>['payload']) => {
+    const extension = extname(fileKey)
+    const needsTemplateUpdate = ['.liquid', '.json'].includes(extension)
+    const isAsset = fileKey.startsWith('assets/')
+
+    if (needsTemplateUpdate && !isAsset) {
+      contentPromise
+        .then((content) => {
+          inMemoryTemplates[fileKey] = content
+          if (extension === '.json') parsedJsonTemplates[fileKey] = JSON.parse(content)
```

Comment:
We need to use `parseJSON` (from `'@shopify/theme-check-node'`) here instead of `JSON.parse` because `templates/*` support JSON with comments and trailing commas.

Reply:
Fixed in https://github.com/Shopify/cli/pull/4302/commits/dddf9a19fcc30264b2534982112d95d256d47e2f